guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+        Timeout { inner, timeout }
+    }
+}
+```
+
+Note that we don't have to add any trait bounds on `S` even though we expect `S`
+to implement `Service`. In Rust is common practice to only add trait bounds
+where they're required and since `new` doesn't require anything from `S` we
+don't add any trait bounds.
+
+Now the interesting bit. How to implement `Service` for `Timeout<S>`? Lets start
+with an implementation that just forwards everything to the inner service:
+
+```rust
+use tower::Service;
+use std::task::{Context, Poll};
+
+impl<S, Request> Service<Request> for Timeout<S>
+where
+    S: Service<Request>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Our middleware doesn't care about backpressure so its ready as long
+        // as the inner service is ready.
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request)
+    }
+}
+```
+
+Until you've written lots of middleware writing out a skeleton like this makes
+the process a bit easier.
+
+To actually add a timeout to the inner service what we essentially have to do is
+detect when the future returned by `self.inner.call(request)` has been running
+longer than `self.duration` and abort with an error.
+
+The approach we're going to take is to call [`tokio::time::sleep`] to get a
+future that completes when we're out of time and then select the value from
+whichever of the two futures is the first to complete. We could also use
+`tokio::time::timeout` but `sleep` works just as well.
+
+Creating both futures is done like this:
+
+```rust
+use tokio::time::sleep;
+
+fn call(&mut self, request: Request) -> Self::Future {
+    let response_future = self.inner.call(request);
+
+    // This variable has type `tokio::time::Sleep`.
+    //
+    // We don't have to clone `self.duration` as it implements the `Copy` trait.
+    let sleep = tokio::time::sleep(self.timeout);
+
+    // what to write here?
+}
+```
+
+One possible return type is `Pin<Box<dyn Future<...>>>`. However we want our
+`Timeout` to add as little overhead as possible, so we would like to find a way
+to avoid allocating a `Box`. Imagine we have a large stack, with dozens of
+nested `Service`s, where each layer allocates a new `Box` for every request
+that passes through it. That would result in a lot of allocations which might
+impact performance.
+
+## The response future
+
+To avoid using `Box` lets instead write our own `Future` implementation. We
+start by creating a struct called `ResponseFuture`. It has to be generic over
+the inner service's response future type. This is analogous to wrapping services
+in other services, but this time we're wrapping futures in other futures.
+
+```rust
+use tokio::time::Sleep;
+
+pub struct ResponseFuture<F> {
```

Comment:
```suggestion
pub struct FutureWithTimeout<F> {
```

Reply:
I actually think I'm going to stick with `ResponseFuture` as that is what its called in Tower. I quite like that people can read this guide and when they're done be familiar with the code in Tower.

Reply:
Yeah, I suppose it's okay in Tower it's okay as the type lives in its own module.